warehouse/query-core/src/main/java/datawave/query/transformer/UniqueTransform.java

```diff
@@ -0,0 +1,168 @@
+    
+    public UniqueTransform(Set<String> fields) {
+        this.fields = fields;
+        this.bloom = BloomFilter.create(new ByteFunnel(), 500000, 1e-15);
+        if (DEBUG) {
+            this.seen = new HashSet<ByteSequence>();
+        }
+        if (log.isTraceEnabled())
+            log.trace("unique fields: " + this.fields);
+    }
+    
+    /**
+     * If passing the logic in, then the model being used by the logic then capture the reverse field mapping
+     *
+     * @param logic
+     * @param fields
+     */
+    public UniqueTransform(BaseQueryLogic<Entry<Key,Value>> logic, Set<String> fields) {
+        this(fields);
+        QueryModel model = ((ShardQueryLogic) logic).getQueryModel();
+        if (model != null) {
+            modelMapping = HashMultimap.create();
+            // reverse the reverse query mapping which will give us a mapping from the final field name to the original field name(s)
+            for (Map.Entry<String,String> entry : model.getReverseQueryMapping().entrySet()) {
+                modelMapping.put(entry.getValue(), entry.getKey());
+            }
+        }
+    }
+    
+    public Predicate<Entry<Key,Document>> getUniquePredicate() {
+        return new Predicate<Entry<Key,Document>>() {
+            @Override
+            public boolean apply(@Nullable Entry<Key,Document> input) {
+                return UniqueTransform.this.apply(input) != null;
+            }
+        };
+    }
+    
+    @Nullable
+    @Override
+    public Entry<Key,Document> apply(@Nullable Entry<Key,Document> keyDocumentEntry) {
+        if (keyDocumentEntry != null) {
+            try {
+                if (isDuplicate(keyDocumentEntry.getValue())) {
+                    keyDocumentEntry = null;
+                }
+            } catch (IOException ioe) {
+                log.error("Failed to convert document to bytes.  Returning document as unique.", ioe);
+            }
+        }
+        return keyDocumentEntry;
+    }
+    
+    private byte[] getBytes(Document document) throws IOException {
+        // we need to pull the fields out of the document.
+        ByteArrayOutputStream bytes = new ByteArrayOutputStream();
+        DataOutputStream output = new DataOutputStream(bytes);
+        for (String field : fields) {
+            output.writeChars(field);
+            List<Attribute<?>> attrs = getFields(document, field);
+            if (!attrs.isEmpty()) {
+                for (Attribute<?> attr : attrs) {
+                    attr.write(output, true);
+                }
+            }
+        }
+        output.flush();
+        return bytes.toByteArray();
+    }
+    
+    private List<Attribute<?>> getFields(Document document, String field) {
+        List<Attribute<?>> attrs = new ArrayList<>();
+        List<String> fields = new ArrayList<>();
+        if (modelMapping != null) {
+            fields.addAll(modelMapping.get(field));
+        }
+        if (fields.isEmpty()) {
+            fields.add(field);
```

Comment:
shouldn't we add the original field as long as it isn't already in the list regardless of it if is in the model?

Reply:
I don't think you want to add the original field if the list is non-empty.  That implies that the original field was a model field, which wouldn't be present in your document.

Reply:
that assumes that the model entry always returns itself otherwise it would drop the original field. If we are going to do that it should be well documented.

Reply:
Ok, you're right...  Whatever...